File: BraintreePayPal/BTPayPalAccountNonce.m
#import "BTPayPalAccountNonce_Internal.h"

@interface BTPayPalAccountNonce ()
@property (nonatomic, readwrite, copy) NSString *email;
@property (nonatomic, readwrite, copy) NSString *firstName;
@property (nonatomic, readwrite, copy) NSString *lastName;
@property (nonatomic, readwrite, copy) NSString *phone;
@property (nonatomic, readwrite, strong) BTPostalAddress *billingAddress;
@property (nonatomic, readwrite, strong) BTPostalAddress *shippingAddress;
@property (nonatomic, readwrite, copy) NSString *clientMetadataId;
@property (nonatomic, readwrite, copy) NSString *payerId;
@end

@implementation BTPayPalAccountNonce

@synthesize nonce = _nonce;
@synthesize localizedDescription = _localizedDescription;
@synthesize type = _type;

- (instancetype)initWithNonce:(NSString *)nonce
                  description:(NSString *)description
                        email:(NSString *)email
                    firstName:(NSString *)firstName
                     lastName:(NSString *)lastName
                        phone:(NSString *)phone
               billingAddress:(BTPostalAddress *)billingAddress
              shippingAddress:(BTPostalAddress *)shippingAddress
             clientMetadataId:(NSString *)clientMetadataId
                      payerId:(NSString *)payerId
{
    if (self = [super init]) {
        _nonce = nonce;
        _localizedDescription = description;
        _email = email;
        _firstName = firstName;
        _lastName = lastName;
        _phone = phone;
        _billingAddress = [billingAddress copy];
        _shippingAddress = [shippingAddress copy];
        _clientMetadataId = clientMetadataId;
        _payerId = payerId;
        _type = @"PayPal";
    }
    return self;
}

@end

File: Sources/BraintreePayPal/BTPayPalAccountNonce.swift
import Foundation

#if canImport(BraintreeCore)
import BraintreeCore
#endif

/// Contains information about a PayPal payment method
@objcMembers public class BTPayPalAccountNonce: BTPaymentMethodNonce {
    
    /// Payer's email address.
    public let email: String?
    
    /// Payer's first name.
    public let firstName: String?
    
    /// Payer's last name.
    public let lastName: String?
    
    /// Payer's phone number.
    public let phone: String?
    
    /// The billing address.
    public let billingAddress: BTPostalAddress?
    
    /// The shipping address.
    public let shippingAddress: BTPostalAddress?
    
    /// Client metadata id associated with this transaction.
    public let clientMetadataID: String?
    
    /// Optional. Payer id associated with this transaction.
    /// Will be provided for Vault and Checkout.
    public let payerID: String?
    
    /// Optional. Credit financing details if the customer pays with PayPal Credit.
    /// Will be provided for Vault and Checkout.
    public let creditFinancing: BTPayPalCreditFinancing?
    
    init?(json: BTJSON) {
        guard let nonce = json["nonce"].asString() else { return nil }
        
        let details = json["details"]
        let payerInfo = details["payerInfo"]

        self.email = payerInfo["email"].asString() ?? details["email"].asString()
        self.firstName = payerInfo["firstName"].asString()
        self.lastName = payerInfo["lastName"].asString()
        self.phone = payerInfo["phone"].asString()
        self.billingAddress = payerInfo["billingAddress"].asAddress()
        self.shippingAddress = payerInfo["shippingAddress"].asAddress() ?? payerInfo["accountAddress"].asAddress()
        self.clientMetadataID = details["correlationId"].asString()
        self.payerID = payerInfo["payerId"].asString()
        self.creditFinancing = details["creditFinancingOffered"].asPayPalCreditFinancing()
        super.init(nonce: nonce, type: "PayPal", isDefault: json["default"].isTrue)
    }
}
